crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -367,136 +404,184 @@ pub(super) struct FlatAbi {
-    /// return the result via the specified `Sender`.
-    Write {
-        buffer: B,
-        tx: oneshot::Sender<HostResult<B>>,
-    },
-    /// Drop the write end of the stream or future.
-    Drop(Option<Box<dyn FnOnce() -> B + Send + Sync>>),
-    /// Watch the read (i.e. opposite) end of this stream or future, dropping
-    /// the specified sender when it is dropped.
-    Watch { tx: oneshot::Sender<()> },
+/// Trait representing objects (such as streams, futures, or structs containing
+/// them) which require access to the store in order to be disposed of properly.
+pub trait DropWithStore: Sized {
+    /// Dispose of `self` using the specified store.
+    fn drop(self, store: impl AsContextMut) -> Result<()>;
+
+    /// Dispose of `self` using the specified accessor.
+    fn drop_with(self, accessor: impl AsAccessor) -> Result<()> {
+        accessor.as_accessor().with(move |store| self.drop(store))
+    }
 }
 
-/// Represents a pending event on a host-owned read end of a stream or future.
-///
-/// See `ComponentInstance::start_read_event_loop` for details.
-enum ReadEvent<B> {
-    /// Read as many items as the specified buffer will hold from the stream or
-    /// future, and return the result via the specified `Sender`.
-    Read {
-        buffer: B,
-        tx: oneshot::Sender<HostResult<B>>,
-    },
-    /// Drop the read end of the stream or future.
-    Drop,
-    /// Watch the write (i.e. opposite) end of this stream or future, dropping
-    /// the specified sender when it is dropped.
-    Watch { tx: oneshot::Sender<()> },
+/// Trait representing objects (such as the writable end of a future or a struct
+/// containing one) which require access to the store _and_ a value to write in
+/// order to be disposed of properly.
+pub trait DropWithStoreAndValue<T: func::Lower + Send + Sync + 'static>: Sized {
+    /// Dispose of `self` using the specified store, writing the specified value.
+    fn drop(self, store: impl AsContextMut, value: T) -> Result<()>;
+
+    /// Dispose of `self` using the specified accessor and value.
+    fn drop_with(self, accessor: impl AsAccessor, value: T) -> Result<()> {
+        accessor
+            .as_accessor()
+            .with(move |store| self.drop(store, value))
+    }
 }
 
-/// Send the specified value to the specified `Sender`.
+/// RAII wrapper for `DropWithStore` implementations.
 ///
-/// This will panic if there is no room in the channel's buffer, so it should
-/// only be used in a context where there is at least one empty spot in the
-/// buffer.  It will silently ignore any other error (e.g. if the `Receiver` has
-/// been dropped).
-fn send<T>(tx: &mut mpsc::Sender<T>, value: T) {
-    if let Err(e) = tx.try_send(value) {
-        if e.is_full() {
-            unreachable!();
+/// This may be used to automatically dispose of the wrapped object when it goes
+/// out of scope.
+pub struct WithAccessor<'a, T: DropWithStore, U: 'static, D: HasData = HasSelf<U>> {
```

Comment:
Bikeshedding this a bit, this feels like it should be called `DropWithAccessor` or somehow have "drop" related to the name since "with accessor" is a super generic name that doesn't quite imply what's happening under the hood here. Not that I like the name `DropWithAccessor` though, that also feels bad...